### ISSUE TITLE
The loss during the training process of DSB2018 is "nan".

### ISSUE THREAD

**Issue by jinzhengbei** [tpnam0901/QTSeg#14]
During the training process of the DSB2018 dataset, the loss became nan. Starting from a certain epoch, all the losses changed to nan. Other data and everything are normal. Could it be due to the data set? Thank you.

Train epoch 4/350
2025-08-08 07:09:41,067 - Train epoch 4/350
Epoch 4/350 - loss: 0.3805 - mae: 4.3458 - acc: 95.6542 - dice: 79.0254 - iou: 67
Epoch 4/350 - epoch_loss: 0.4162 - mae: 5.3200 - acc: 94.6800 - dice: 77.5373 - iou: 65.6791
2025-08-08 07:09:59,941 - Epoch 4/350 - epoch_loss: 0.4162 - mae: 5.3200 - acc: 94.6800 - dice: 77.5373 - iou: 65.6791
Validation epoch 4/350
2025-08-08 07:10:00,296 - Validation epoch 4/350
100%|██████████████████████████████████████████| 134/134 [00:15<00:00,  8.82it/s]
Epoch 4/350 - val_loss: 0.3721 - mae: 4.9237 - acc: 95.0763 - dice: 79.3424 - iou: 68.4110
2025-08-08 07:10:15,567 - Epoch 4/350 - val_loss: 0.3721 - mae: 4.9237 - acc: 95.0763 - dice: 79.3424 - iou: 68.4110
Loss improved from 0.3951 to 0.3721
2025-08-08 07:10:15,567 - Loss improved from 0.3951 to 0.3721
Metric mae improved from 5.2210 to 4.9237
2025-08-08 07:10:15,718 - Metric mae improved from 5.2210 to 4.9237
Metric acc improved from 94.7790 to 95.0763
2025-08-08 07:10:15,896 - Metric acc improved from 94.7790 to 95.0763
Metric dice improved from 78.0509 to 79.3424
2025-08-08 07:10:16,049 - Metric dice improved from 78.0509 to 79.3424
Metric iou improved from 66.5973 to 68.4110
2025-08-08 07:10:16,245 - Metric iou improved from 66.5973 to 68.4110
Number of trainable parameters: 6955314
2025-08-08 07:10:16,483 - Number of trainable parameters: 6955314

Train epoch 5/350
2025-08-08 07:10:16,483 - Train epoch 5/350
Epoch 5/350 - loss: nan - mae: 13.6832 - acc: 86.3169 - dice: 0.0112 - iou: 0.005
Epoch 5/350 - epoch_loss: nan - mae: 8.6355 - acc: 91.3645 - dice: 39.1503 - iou: 33.3692
2025-08-08 07:10:36,422 - Epoch 5/350 - epoch_loss: nan - mae: 8.6355 - acc: 91.3645 - dice: 39.1503 - iou: 33.3692
Validation epoch 5/350
2025-08-08 07:10:37,207 - Validation epoch 5/350
100%|██████████████████████████████████████████| 134/134 [00:30<00:00,  4.44it/s]
Epoch 5/350 - val_loss: nan - mae: 11.8622 - acc: 88.1378 - dice: 0.0200 - iou: 0.0100
2025-08-08 07:11:07,584 - Epoch 5/350 - val_loss: nan - mae: 11.8622 - acc: 88.1378 - dice: 0.0200 - iou: 0.0100
Number of trainable parameters: 6955314
2025-08-08 07:11:07,587 - Number of trainable parameters: 6955314
Train epoch 6/350
2025-08-08 07:11:07,587 - Train epoch 6/350
Epoch 6/350 - loss: nan - mae: 10.8446 - acc: 89.1554 - dice: 0.0188 - iou: 0.009
Epoch 6/350 - epoch_loss: nan - mae: 11.9259 - acc: 88.0741 - dice: 0.2197 - iou: 0.2075
2025-08-08 07:11:38,472 - Epoch 6/350 - epoch_loss: nan - mae: 11.9259 - acc: 88.0741 - dice: 0.2197 - iou: 0.2075
Validation epoch 6/350
2025-08-08 07:11:39,099 - Validation epoch 6/350
100%|██████████████████████████████████████████| 134/134 [00:10<00:00, 13.29it/s]
Epoch 6/350 - val_loss: nan - mae: 11.8622 - acc: 88.1378 - dice: 0.0200 - iou: 0.0100
2025-08-08 07:11:49,294 - Epoch 6/350 - val_loss: nan - mae: 11.8622 - acc: 88.1378 - dice: 0.0200 - iou: 0.0100
Number of trainable parameters: 6955314
2025-08-08 07:11:49,297 - Number of trainable parameters: 6955314

**Comment by tpnam0901** [comment]
Hi @jinzhengbei, could you share the dataset after preprocessing and your config for training?

**Comment by jinzhengbei** [comment]
``from configs.base import Config as BaseConfig


class Config(BaseConfig):
    # Base
    def __init__(self, **kwargs):
        super(Config, self).__init__(**kwargs)
        self.add_args()
        for key, value in kwargs.items():
            setattr(self, key, value)

    def add_args(self, **kwargs):

        # fold_0, fold_1, fold_2, fold_3, fold_4
        self.data_fold = "fold_1"
        self.data_root: str = f"/raid/jzb/QTSeg/QTSeg/src/working/datasets/DSB2018/folds/{self.data_fold}"
        self.dataloader: str = "DSB2018A"
        self.valid_type: str = "val"
        self.cvtColor = None

        self.name = self.model_type + "/" + self.encoder_model + self.decoder_model

        for key, value in kwargs.items():
            setattr(self, key, value)

<img width="285" height="366" alt="Image" src="https://github.com/user-attachments/assets/b45123d4-a9db-46b3-a75e-5482bface8ee" />

<img width="975" height="900" alt="Image" src="https://github.com/user-attachments/assets/071c8ae3-3a7c-4003-aef4-9800db618771" />

These are the config and dataset. I followed your scripts, combined and splited. But i do not know why.

**Comment by tpnam0901** [comment]
Hi @jinzhengbei 
The reason for getting nan in the loss function is because of the ```augment_seg``` function in [dataset.py](https://github.com/tpnam0901/QTSeg/blob/main/src/data/dataset.py)

It makes some input images become a zero tensor. 

You can solve this problem by choosing a better SEED value, or you can replace the ```augment_seg``` function with this function

```python
def augment_seg(self, img, seg):
        seg_raw = seg
        seg = self.mask_to_onehot(seg)
        aug_det = self.img_aug.to_deterministic()
        image_aug = aug_det.augment_image(img)

        # This line will fix the problem
        if len(np.unique(image_aug)) == 1:
            return img, seg_raw

        segmap = ia.SegmentationMapOnImage(
            seg, nb_classes=np.max(seg) + 1, shape=img.shape
        )
        segmap_aug = aug_det.augment_segmentation_maps(segmap)
        segmap_aug = segmap_aug.get_arr_int()
        segmap_aug = np.argmax(segmap_aug, axis=-1).astype(np.float32)
        return image_aug, segmap_aug
```
